modules/audio_coding/codecs/opus/audio_encoder_opus.cc

```diff
@@ -623,18 +618,189 @@ AudioEncoder::EncodedInfo AudioEncoderOpusImpl::EncodeImpl(
   info.encoded_timestamp = first_timestamp_in_buffer_;
   info.payload_type = payload_type_;
   info.send_even_if_empty = true;  // Allows Opus to send empty packets.
+
   // After 20 DTX frames (MAX_CONSECUTIVE_DTX) Opus will send a frame
   // coding the background noise. Avoid flagging this frame as speech
   // (even though there is a probability of the frame being speech).
-  info.speech = !dtx_frame && (consecutive_dtx_frames_ != 20);
+  info.speech = IsPacketSpeech(info.encoded_bytes, encoded->data());
+
   info.encoder_type = CodecType::kOpus;
 
-  // Increase or reset DTX counter.
-  consecutive_dtx_frames_ = (dtx_frame) ? (consecutive_dtx_frames_ + 1) : (0);
+  // Will use new packet size for next encoding.
+  config_.frame_size_ms = next_frame_length_ms_;
 
   return info;
 }
 
+// RingRTC change to detect if an encoded packet should be considered to contain
+// speech or not. Generally, if the last frame in the packet is audio, it is
+// speech, unless it is a DTX refresh frame. This function follows RFC-6716 to
+// check frames in each encoded packet.
+bool AudioEncoderOpusImpl::IsPacketSpeech(
+    size_t encoded_bytes,
+    uint8_t* encoded) {
+  bool speech = false;
+
+  // If the encoder returns 0, 1, or 2 encoded bytes, by definition, the packet
+  // contains only DTX frame(s). 0 is the special case in which
+  // opus_interface.cc detected consecutive DTX packets and is instructing
+  // WebRTC not to send any packet out over the wire.
+  bool dtx_packet = (encoded_bytes <= 2);
+
+  if (config_.frame_size_ms > 20) {
+    // For packet times greater than 20ms, Opus will encode a group of 20ms
+    // frames and combine them into a 'packet' with a TOC.
+    if (dtx_packet) {
+      // The 'packet' contains only DTX frames.
+      consecutive_dtx_frames_ += config_.frame_size_ms / 20;
+    } else {
+      // The 'packet' contains at least one non-DTX frame.
+      if (((encoded[0] & 0x98) == 0x08) ||
+          ((encoded[0] & 0x78) == 0x78) ||
+          ((encoded[0] & 0x98) == 0x98)) {
```

Comment:
We only care about packets containing 20ms data as per the [RFC](https://datatracker.ietf.org/doc/html/rfc6716#section-3.1). This check should match 1, 5, 9, and 13, then 15, then 19, 23, 27, and 31.